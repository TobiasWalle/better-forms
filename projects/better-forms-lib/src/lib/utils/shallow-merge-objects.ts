export function shallowMergeObjects<T>(objects: T[]): T {
  return objects.reduce((object, nextObject) => {
    return {
      ...object,
      ...nextObject
    };
  }, {});
}
