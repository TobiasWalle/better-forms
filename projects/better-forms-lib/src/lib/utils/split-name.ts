export function splitName(path: string): string[] {
  return path.split('.').filter(s => s !== '');
}
