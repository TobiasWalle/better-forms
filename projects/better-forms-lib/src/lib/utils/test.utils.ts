import { ControlValueAccessor } from '@angular/forms';
import { Observable } from 'rxjs';
import { take } from 'rxjs/operators';

export function createMockValueAccessor(): ControlValueAccessor {
  return {
    writeValue: jest.fn(),
    registerOnChange: jest.fn(),
    registerOnTouched: jest.fn(),
    setDisabledState: jest.fn(),
  };
}

export async function getLatestValue<T>(observable: Observable<T>): Promise<T> {
  return await observable.pipe(take(1)).toPromise();
}
