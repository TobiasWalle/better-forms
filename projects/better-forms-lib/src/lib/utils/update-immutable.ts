const immutable: any = require('object-path-immutable');

export function setImmutable<T>(original: T, path: string[], newValue: any): T {
  return immutable.set(original, path, newValue);
}

export function getByPath<T>(object: T, path: string[]): any {
  return path.reduce((nextObject: any, key: string) => {
    if (key in nextObject) {
      return nextObject[key];
    }
    throw new Error(`Couldn't find "${path}" on object. "${key}" doesn not exists on ${nextObject}.`);
  }, object);
}

