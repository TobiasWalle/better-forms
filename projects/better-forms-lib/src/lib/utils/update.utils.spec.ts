import { getByPath, setImmutable } from './update.utils';

describe('setImmutable', () => {
  it('should update an object', () => {
    const object = Object.freeze({
      a: {
        b: 1
      },
      c: 2
    });

    const newObject = setImmutable(object, ['a', 'b'], 2);

    expect(object).toEqual(({
      a: {
        b: 1
      },
      c: 2
    }));
    expect(newObject).toEqual({
      a: {
        b: 2
      },
      c: 2
    });
  });

  it('should update an path that is not defined', () => {
    const object = Object.freeze({
      a: {
        b: 1
      },
      c: 2
    });

    const newObject = setImmutable(object, ['d'], 3);

    expect(newObject).toEqual({
      a: {
        b: 1
      },
      c: 2,
      d: 3
    });
  });

  it('should update arrays', () => {
    const object = Object.freeze({
      a: [],
    });

    let newObject = setImmutable(object, ['a', '0'], 2);

    expect(newObject).toEqual({
      a: [2]
    });

    newObject = setImmutable(object, ['a', '0'], 3);

    expect(newObject).toEqual({
      a: [3]
    });

    newObject = setImmutable(object, ['a', '0'], undefined);

    expect(newObject).toEqual({
      a: [undefined]
    });
  });
});

describe('getByPath', () => {
  it('should get a value by the path', () => {
    expect(getByPath({ a: { b: 1 } }, ['a', 'b'])).toBe(1);
  });

  it('should throw error if the path does not exists', () => {
    expect(() => getByPath({ a: { b: 1 } }, ['a', 'c'])).toThrowErrorMatchingSnapshot();
  });

  it('should not throw error if the path is undefined', () => {
    expect(getByPath({ a: { b: undefined } }, ['a', 'b'])).toBe(undefined);
  });
});
