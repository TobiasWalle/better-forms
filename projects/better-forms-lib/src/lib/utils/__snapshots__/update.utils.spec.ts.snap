// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getByPath should throw error if the path does not exists 1`] = `
"Couldn't find \\"a,c\\" on object. \\"c\\" does not exists on {
  \\"b\\": 1
}."
`;
